### PR TITLE
Introduce a token and parsing phase to the replay parser

### PR DESCRIPTION
**Pull request by Garanas**
The first step towards extracting more information from a replay is separating the various phases of the parsing. With these changes we separate the following phases for the body of the replay:

- (1) Token phase
- (2) Parsing phase
- (3) Semantics (interpretation) phase

Now it is much less error prone when we want to read and interpret more data from the replay.

As a nice side product, the body of the replay is now also nicely documented by the various (sealed) interfaces and the records implementing them.

With thanks to @Sheikah45 for the pair programming sessions

### CODE REVIEW

**Comment by Garanas** [comment]
The next phase would be to introduce a similar way to process semantics. At the moment the entire replay class is tightly coupled - but what one would want to do is simply apply semantics to the current list of parsed tokens, which returns (one or more) lists of parsed tokens again. That way you can compose the functionality.

**Review by Sheikah45 — APPROVED** [review]

File: faf-commons-data/src/main/java/com/faforever/commons/replay/event/Event.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+    NONE("NONE"),
+    STOP("Stop"),
+    MOVE("Move"),
+    DIVE("Dive"),
+    FORM_MOVE("FormMove"),
+    BUILD_SILO_TACTICAL("BuildSiloTactical"),
+    BUILD_SILO_NUKE("BuildSiloNuke"),
+    BUILD_FACTORY("BuildFactory"),
+    BUILD_MOBILE("BuildMobile"),
+    BUILD_ASSIsT("BuildAssist"),
```
Comment (Brutus5000) [inline faf-commons-data/src/main/java/com/faforever/commons/replay/event/Event.java]: capitalize all letters (found by lack of syntax highlighting lol)

File: faf-commons-data/src/main/java/com/faforever/commons/replay/event/Event.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+  /**
+   *
+   * @param commandId   The command id to reference it later (to adjust or delete it, for example)
+   * @param commandtype   The command type (move, attack - what else?)
+   * @param commandTarget
+   * @param commandFormation
+   * @param blueprintId
+   * @param parametersLua
+   */
+  record CommandData (int commandId, CommandType commandtype, CommandTarget commandTarget, CommandFormation commandFormation, String blueprintId, LuaData parametersLua) {}
```
Comment (Brutus5000) [inline faf-commons-data/src/main/java/com/faforever/commons/replay/event/Event.java]: should be `commandType` (also in javadoc)

**Comment by Garanas** [comment]
Running it in parallel allowed the parsing to use more of my systems resources. The determinism (order of tokens) is not lost, as the test suite shows.

**Review by Brutus5000 — COMMENTED** [review]
Well looks I'm late for the party

File: faf-commons-data/src/main/java/com/faforever/commons/replay/event/Event.java (hunk 2 above)
Comment (Garanas) [reply faf-commons-data/src/main/java/com/faforever/commons/replay/event/Event.java]: I'll pick these up when I do the header of the replay 👍

**Pull request merged**